Create rule S6832: Non-singleton Spring beans should not be injected in a Singleton bean (#3355)
SonarSource/rspec: rules/S6832/java/rule.adoc

In Spring Boot the scope of a bean defines the lifecycle and visibility of that bean in the Spring container.
There are six scopes:

- *Singleton*: default, one instance per Spring container
- *Prototype*: a new instance per bean request
- *Request*: a new instance per HTTP request
- *Session*: a new instance per HTTP session
- *Application*: a new instance per ServletContext
- *Websocket*: a new instance per Websocket session
The last four scopes mentioned, request, session, application and websocket, are only available in a web-aware application.

== Why is this an issue?

> In Spring Boot, Singleton beans and their dependencies are initialized when the application context is created.

Reviewer: I believe `Singleton` should not be capitalized, as it is not a proper name.

If a Singleton bean depends on a bean with a shorter-lived scope (like a Request or Session scoped bean),
it retains the same instance of that bean, even when new instances are created for each Request or Session.
This mismatch can cause unexpected behavior and bugs, as the Singleton bean doesn't interact correctly with the new instances of the shorter-lived bean.

> This rule raises an issue for the injection of non-Singleton beans in a Singleton bean.

Reviewer: ```suggestion
This rule raises an issue when non-singleton beans are injected into a singleton bean.
```

=== What is the potential impact?

When a Singleton bean has a dependency on a bean with a shorter-lived scope, it can lead to the following issues:

- *Data inconsistency*: any state change in the shorter-lived bean will not be reflected in the Singleton bean.

- *Incorrect behavior*: using the same instance of the shorter-lived bean, when a new instance is supposed to be created for each new request or session.

> - *Memory leaks*: preventing garbage collection of a shorter-lived bean that allocates a significant amount of data over time.

Reviewer: Nice! We rarely have this section.

== How to fix it

Inject a shorter-lived bean into a Singleton bean using *ApplicationContext*, *Factories* or *Providers*.

=== Code examples

==== Noncompliant code example

When a Singleton bean auto-wires a Request scoped bean, the dependency is resolved at instantiation time and thus the same instance is used for each HTTP request.

[source,java,diff-id=1,diff-type=noncompliant]
----
@Component
@Scope(value = WebApplicationContext.SCOPE_REQUEST, proxyMode = ScopedProxyMode.TARGET_CLASS)
public class RequestBean {
    //...
}
public class SingletonBean {
    @Autowired

>     private final RequestBean requestBeanFactory; // Noncompliant, the same instance of RequestBean is used for each HTTP request.

Reviewer: ```suggestion
    private final RequestBean requestBean; // Noncompliant, the same instance of RequestBean is used for each HTTP request.
```

    public RequestBean getRequestBean() {

>         return requestBeanFactory;

Reviewer: ```suggestion
        return requestBean;
```

    }
}
----

==== Compliant solution

> Using as injection point a `ObjectFactory<RequestBean>`, s `ObjectProvider<RequestBean>` or a `Provider<RequestBean>` (as for JSR-330).

Reviewer: ```suggestion
Instead use a `ObjectFactory<RequestBean>`, `ObjectProvider<RequestBean>`, or `Provider<RequestBean>` as injection point (as for JSR-330).
```

> Using as injection point a `ObjectFactory<RequestBean>`, s `ObjectProvider<RequestBean>` or a `Provider<RequestBean>` (as for JSR-330).

Reviewer: Can we add a clickable link to what JSR-330 is?
